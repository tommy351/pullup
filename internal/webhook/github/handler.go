package github

import (
	"context"
	"errors"
	"fmt"
	"net/http"

	"github.com/go-logr/logr"
	"github.com/google/go-github/v32/github"
	"github.com/google/wire"
	"github.com/tommy351/pullup/internal/httputil"
	"github.com/tommy351/pullup/internal/webhook/hookutil"
	"github.com/tommy351/pullup/pkg/apis/pullup/v1alpha1"
	"github.com/tommy351/pullup/pkg/apis/pullup/v1beta1"
	"k8s.io/apimachinery/pkg/runtime"
	"k8s.io/client-go/tools/record"
	"sigs.k8s.io/controller-runtime/pkg/client"
	"sigs.k8s.io/controller-runtime/pkg/manager"
)

// +kubebuilder:rbac:groups=pullup.dev,resources=webhooks,verbs=get;list;watch
// +kubebuilder:rbac:groups=pullup.dev,resources=githubwebhooks,verbs=get;list;watch
// +kubebuilder:rbac:groups=pullup.dev,resources=resourcesets,verbs=create;patch;delete

const (
	nameField      = "spec.repositories.githubName"
	repoTypeGitHub = "github"
)

// HandlerConfigSet provides a handler config.
// nolint: gochecknoglobals
var HandlerConfigSet = wire.NewSet(
	wire.Struct(new(HandlerConfig), "*"),
)

// HandlerSet provides a handler.
// nolint: gochecknoglobals
var HandlerSet = wire.NewSet(
	HandlerConfigSet,
	NewHandler,
)

type Config struct {
	Secret string `mapstructure:"secret"`
}

type HandlerConfig struct {
	Config                 Config
	Client                 client.Client
	Recorder               record.EventRecorder
	ResourceTemplateClient hookutil.ResourceTemplateClient
}

type Handler struct {
	HandlerConfig
}

func NewHandler(conf HandlerConfig, mgr manager.Manager) (*Handler, error) {
	indexer := mgr.GetFieldIndexer()
	err := indexer.IndexField(context.TODO(), &v1alpha1.Webhook{}, nameField, func(obj runtime.Object) []string {
		var result []string

		for _, repo := range obj.(*v1alpha1.Webhook).Spec.Repositories {
			if repo.Type == repoTypeGitHub {
				result = append(result, repo.Name)
			}
		}

		return result
	})
	if err != nil {
		return nil, fmt.Errorf("index failed: %w", err)
	}

	err = indexer.IndexField(context.TODO(), &v1beta1.GitHubWebhook{}, nameField, func(obj runtime.Object) []string {
		var result []string

		for _, repo := range obj.(*v1beta1.GitHubWebhook).Spec.Repositories {
			result = append(result, repo.Name)
		}

		return result
	})
	if err != nil {
		return nil, fmt.Errorf("index failed: %w", err)
	}

	return &Handler{
		HandlerConfig: conf,
	}, nil
}

func (h *Handler) Handle(w http.ResponseWriter, r *http.Request) error {
	logger := logr.FromContextOrDiscard(r.Context())
	payload, err := h.parsePayload(r)
	if err != nil {
		logger.Error(err, "Invalid payload")

		return httputil.JSON(w, http.StatusBadRequest, &httputil.Response{
			Errors: []httputil.Error{
				{Description: "Invalid payload"},
			},
		})
	}

	if err := h.handlePayload(r, payload); err != nil {
		if errors.Is(err, hookutil.ErrResourceNameRequired) {
			return httputil.JSON(w, http.StatusBadRequest, &httputil.Response{
				Errors: []httputil.Error{
					{Description: "resourceName is not set in the webhook"},
				},
			})
		}

		return err
	}

	return httputil.JSON(w, http.StatusOK, &httputil.Response{})
}

func (h *Handler) handlePayload(r *http.Request, payload interface{}) error {
	switch event := payload.(type) {
	case *github.PushEvent:
		return h.handlePushEvent(r.Context(), event)
	case *github.PullRequestEvent:
		return h.handlePullRequestEvent(r.Context(), event)
	}

	return nil
}

func (h *Handler) parsePayload(r *http.Request) (interface{}, error) {
	payload, err := github.ValidatePayload(r, []byte(h.Config.Secret))
	if err != nil {
		return nil, fmt.Errorf("invalid github payload: %w", err)
	}

	return github.ParseWebHook(github.WebHookType(r), payload)
}
