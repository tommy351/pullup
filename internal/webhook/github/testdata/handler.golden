# Generated by goldga. DO NOT EDIT.
[snapshots]
"Handler v1alpha1.Webhook when event type is pull request when action = opened when resource set does not exist should match the golden file" = '''
- apiVersion: pullup.dev/v1alpha1
  kind: ResourceSet
  metadata:
    creationTimestamp: null
    labels:
      pull-request-number: "46"
      webhook-name: foobar
    name: foobar-46
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: Webhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1alpha1/namespaces/test/resourcesets/foobar-46
  spec:
    base:
      ref: base
      sha: b436f6eb3356504235c0c9a8e74605c820d8d9cc
    head:
      ref: test
      sha: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
    number: 46
    resources:
    - apiVersion: v1
      kind: Pod
      metadata:
        name: foo
  status: {}
'''
"Handler v1alpha1.Webhook when event type is pull request when action = opened when resource set exists should match the golden file" = '''
- apiVersion: pullup.dev/v1alpha1
  kind: ResourceSet
  metadata:
    creationTimestamp: null
    labels:
      pull-request-number: "46"
      webhook-name: foobar
    name: foobar-46
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: Webhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1alpha1/namespaces/test/resourcesets/foobar-46
  spec:
    base:
      ref: base
      sha: b436f6eb3356504235c0c9a8e74605c820d8d9cc
    head:
      ref: test
      sha: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
    number: 46
    resources:
    - apiVersion: v1
      kind: Pod
      metadata:
        name: foo
  status: {}
'''
"Handler v1alpha1.Webhook when event type is pull request when action = reopened when resource set does not exist should match the golden file" = '''
- apiVersion: pullup.dev/v1alpha1
  kind: ResourceSet
  metadata:
    creationTimestamp: null
    labels:
      pull-request-number: "46"
      webhook-name: foobar
    name: foobar-46
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: Webhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1alpha1/namespaces/test/resourcesets/foobar-46
  spec:
    base:
      ref: base
      sha: b436f6eb3356504235c0c9a8e74605c820d8d9cc
    head:
      ref: test
      sha: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
    number: 46
    resources:
    - apiVersion: v1
      kind: Pod
      metadata:
        name: foo
  status: {}
'''
"Handler v1alpha1.Webhook when event type is pull request when action = reopened when resource set exists should match the golden file" = '''
- apiVersion: pullup.dev/v1alpha1
  kind: ResourceSet
  metadata:
    creationTimestamp: null
    labels:
      pull-request-number: "46"
      webhook-name: foobar
    name: foobar-46
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: Webhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1alpha1/namespaces/test/resourcesets/foobar-46
  spec:
    base:
      ref: base
      sha: b436f6eb3356504235c0c9a8e74605c820d8d9cc
    head:
      ref: test
      sha: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
    number: 46
    resources:
    - apiVersion: v1
      kind: Pod
      metadata:
        name: foo
  status: {}
'''
"Handler v1alpha1.Webhook when event type is pull request when action = synchronize when resource set does not exist should match the golden file" = '''
- apiVersion: pullup.dev/v1alpha1
  kind: ResourceSet
  metadata:
    creationTimestamp: null
    labels:
      pull-request-number: "46"
      webhook-name: foobar
    name: foobar-46
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: Webhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1alpha1/namespaces/test/resourcesets/foobar-46
  spec:
    base:
      ref: base
      sha: b436f6eb3356504235c0c9a8e74605c820d8d9cc
    head:
      ref: test
      sha: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
    number: 46
    resources:
    - apiVersion: v1
      kind: Pod
      metadata:
        name: foo
  status: {}
'''
"Handler v1alpha1.Webhook when event type is pull request when action = synchronize when resource set exists should match the golden file" = '''
- apiVersion: pullup.dev/v1alpha1
  kind: ResourceSet
  metadata:
    creationTimestamp: null
    labels:
      pull-request-number: "46"
      webhook-name: foobar
    name: foobar-46
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: Webhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1alpha1/namespaces/test/resourcesets/foobar-46
  spec:
    base:
      ref: base
      sha: b436f6eb3356504235c0c9a8e74605c820d8d9cc
    head:
      ref: test
      sha: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
    number: 46
    resources:
    - apiVersion: v1
      kind: Pod
      metadata:
        name: foo
  status: {}
'''
"Handler v1alpha1.Webhook when event type is pull request when resource name is set should match the golden file" = '''
- apiVersion: pullup.dev/v1alpha1
  kind: ResourceSet
  metadata:
    creationTimestamp: null
    labels:
      pull-request-number: "46"
      webhook-name: foobar
    name: foobar-bar-46-0ce4cf0450de14c6555c563fa9d36be67e69aa2f
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: Webhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1alpha1/namespaces/test/resourcesets/foobar-bar-46-0ce4cf0450de14c6555c563fa9d36be67e69aa2f
  spec:
    base:
      ref: base
      sha: b436f6eb3356504235c0c9a8e74605c820d8d9cc
    head:
      ref: test
      sha: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
    number: 46
    resources:
    - apiVersion: v1
      kind: Pod
      metadata:
        name: foo
  status: {}
'''
"Handler v1beta1.GitHubWebhook when event type = pull_request when action = opened when resource does not exist should match the golden file" = '''
- apiVersion: pullup.dev/v1beta1
  kind: ResourceTemplate
  metadata:
    creationTimestamp: null
    name: foobar
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1beta1
      blockOwnerDeletion: true
      controller: true
      kind: GitHubWebhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1beta1/namespaces/test/resourcetemplates/foobar
  spec:
    data:
      event:
        action: opened
        number: 46
        pull_request:
          base:
            ref: base
            sha: b436f6eb3356504235c0c9a8e74605c820d8d9cc
          head:
            ref: test
            sha: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
        repository:
          full_name: foo/bar
          name: bar
          owner:
            login: foo
    webhookRef:
      apiVersion: pullup.dev/v1beta1
      kind: GitHubWebhook
      name: foobar
  status: {}
'''
"Handler v1beta1.GitHubWebhook when event type = pull_request when action = opened when resource exists should match the golden file" = '''
- apiVersion: pullup.dev/v1beta1
  kind: ResourceTemplate
  metadata:
    creationTimestamp: null
    name: foobar
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1beta1
      blockOwnerDeletion: true
      controller: true
      kind: GitHubWebhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1beta1/namespaces/test/resourcetemplates/foobar
  spec:
    data:
      event:
        action: opened
        number: 46
        pull_request:
          base:
            ref: base
            sha: b436f6eb3356504235c0c9a8e74605c820d8d9cc
          head:
            ref: test
            sha: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
        repository:
          full_name: foo/bar
          name: bar
          owner:
            login: foo
    webhookRef:
      apiVersion: pullup.dev/v1beta1
      kind: GitHubWebhook
      name: foobar
  status: {}
'''
"Handler v1beta1.GitHubWebhook when event type = pull_request when action = reopened when resource does not exist should match the golden file" = '''
- apiVersion: pullup.dev/v1beta1
  kind: ResourceTemplate
  metadata:
    creationTimestamp: null
    name: foobar
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1beta1
      blockOwnerDeletion: true
      controller: true
      kind: GitHubWebhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1beta1/namespaces/test/resourcetemplates/foobar
  spec:
    data:
      event:
        action: reopened
        number: 46
        pull_request:
          base:
            ref: base
            sha: b436f6eb3356504235c0c9a8e74605c820d8d9cc
          head:
            ref: test
            sha: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
        repository:
          full_name: foo/bar
          name: bar
          owner:
            login: foo
    webhookRef:
      apiVersion: pullup.dev/v1beta1
      kind: GitHubWebhook
      name: foobar
  status: {}
'''
"Handler v1beta1.GitHubWebhook when event type = pull_request when action = reopened when resource exists should match the golden file" = '''
- apiVersion: pullup.dev/v1beta1
  kind: ResourceTemplate
  metadata:
    creationTimestamp: null
    name: foobar
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1beta1
      blockOwnerDeletion: true
      controller: true
      kind: GitHubWebhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1beta1/namespaces/test/resourcetemplates/foobar
  spec:
    data:
      event:
        action: reopened
        number: 46
        pull_request:
          base:
            ref: base
            sha: b436f6eb3356504235c0c9a8e74605c820d8d9cc
          head:
            ref: test
            sha: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
        repository:
          full_name: foo/bar
          name: bar
          owner:
            login: foo
    webhookRef:
      apiVersion: pullup.dev/v1beta1
      kind: GitHubWebhook
      name: foobar
  status: {}
'''
"Handler v1beta1.GitHubWebhook when event type = pull_request when action = synchronize when resource does not exist should match the golden file" = '''
- apiVersion: pullup.dev/v1beta1
  kind: ResourceTemplate
  metadata:
    creationTimestamp: null
    name: foobar
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1beta1
      blockOwnerDeletion: true
      controller: true
      kind: GitHubWebhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1beta1/namespaces/test/resourcetemplates/foobar
  spec:
    data:
      event:
        action: synchronize
        number: 46
        pull_request:
          base:
            ref: base
            sha: b436f6eb3356504235c0c9a8e74605c820d8d9cc
          head:
            ref: test
            sha: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
        repository:
          full_name: foo/bar
          name: bar
          owner:
            login: foo
    webhookRef:
      apiVersion: pullup.dev/v1beta1
      kind: GitHubWebhook
      name: foobar
  status: {}
'''
"Handler v1beta1.GitHubWebhook when event type = pull_request when action = synchronize when resource exists should match the golden file" = '''
- apiVersion: pullup.dev/v1beta1
  kind: ResourceTemplate
  metadata:
    creationTimestamp: null
    name: foobar
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1beta1
      blockOwnerDeletion: true
      controller: true
      kind: GitHubWebhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1beta1/namespaces/test/resourcetemplates/foobar
  spec:
    data:
      event:
        action: synchronize
        number: 46
        pull_request:
          base:
            ref: base
            sha: b436f6eb3356504235c0c9a8e74605c820d8d9cc
          head:
            ref: test
            sha: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
        repository:
          full_name: foo/bar
          name: bar
          owner:
            login: foo
    webhookRef:
      apiVersion: pullup.dev/v1beta1
      kind: GitHubWebhook
      name: foobar
  status: {}
'''
"Handler v1beta1.GitHubWebhook when event type = pull_request when resourceName is not given should match the golden file" = '''
- apiVersion: pullup.dev/v1beta1
  kind: ResourceTemplate
  metadata:
    creationTimestamp: null
    name: foobar-46
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1beta1
      blockOwnerDeletion: true
      controller: true
      kind: GitHubWebhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1beta1/namespaces/test/resourcetemplates/foobar-46
  spec:
    data:
      event:
        action: opened
        number: 46
        pull_request:
          base:
            ref: base
            sha: b436f6eb3356504235c0c9a8e74605c820d8d9cc
          head:
            ref: test
            sha: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
        repository:
          full_name: foo/bar
          name: bar
          owner:
            login: foo
    webhookRef:
      apiVersion: pullup.dev/v1beta1
      kind: GitHubWebhook
      name: foobar
  status: {}
'''
"Handler v1beta1.GitHubWebhook when event type = push when pushing a branch when resource does not exist should match the golden file" = '''
- apiVersion: pullup.dev/v1beta1
  kind: ResourceTemplate
  metadata:
    creationTimestamp: null
    name: foobar
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1beta1
      blockOwnerDeletion: true
      controller: true
      kind: GitHubWebhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1beta1/namespaces/test/resourcetemplates/foobar
  spec:
    data:
      event:
        head: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
        ref: refs/heads/test
        repository:
          full_name: foo/bar
          name: bar
          owner:
            login: foo
    webhookRef:
      apiVersion: pullup.dev/v1beta1
      kind: GitHubWebhook
      name: foobar
  status: {}
'''
"Handler v1beta1.GitHubWebhook when event type = push when pushing a branch when resource exists should match the golden file" = '''
- apiVersion: pullup.dev/v1beta1
  kind: ResourceTemplate
  metadata:
    creationTimestamp: null
    name: foobar
    namespace: test
    ownerReferences:
    - apiVersion: pullup.dev/v1beta1
      blockOwnerDeletion: true
      controller: true
      kind: GitHubWebhook
      name: foobar
      uid: ""
    selfLink: /apis/pullup.dev/v1beta1/namespaces/test/resourcetemplates/foobar
  spec:
    data:
      event:
        head: 0ce4cf0450de14c6555c563fa9d36be67e69aa2f
        ref: refs/heads/test
        repository:
          full_name: foo/bar
          name: bar
          owner:
            login: foo
    webhookRef:
      apiVersion: pullup.dev/v1beta1
      kind: GitHubWebhook
      name: foobar
  status: {}
'''
